### PR TITLE
Richaudience: Shared memory fix

### PR DESCRIPTION
**Pull request by VeronikaSolovei9**
Adapters receive a shallow copy of a request In MakeBids. It means this request has shared pointer values: site, app and other pointer structures are shared between all adapters in request. In order to modify pointer structures you need to create own copy of it and set it back to request. 
The reason why previous unit tests didn't show it is because `imp.tagid` was empty and result `app/site.keywords` didn't change. 
Please review. @richaudience

### CODE REVIEW

File: adapters/richaudience/richaudience.go [hunk 1]
```diff
@@ -52,10 +52,18 @@ func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapte
 		}
 
 		if request.App != nil {
+
```
Comment (SyntaxNode) [inline adapters/richaudience/richaudience.go:55]: Nitpick: Please remove the leading white space here and for the Site check.

File: adapters/richaudience/richaudience.go [hunk 2]
```diff
@@ -160,18 +168,18 @@ func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.R
...
-					seatBid.Bid[i].H = reqBid.Video.H
+					bidResp.SeatBid[j].Bid[i].W = reqBid.Video.W
+					bidResp.SeatBid[j].Bid[i].H = reqBid.Video.H
 				}
 
 				b := &adapters.TypedBid{
-					Bid:     &seatBid.Bid[i],
+					Bid:     &bidResp.SeatBid[j].Bid[i],
 					BidType: bidType,
 				}
```
Comment (SyntaxNode) [inline adapters/richaudience/richaudience.go:184]: Why are these changes necessary for the bid processing loop? I'm not seeing the issue with the original code.

File: adapters/richaudience/richaudience.go (hunk 1 above)
Comment (VeronikaSolovei9) [reply adapters/richaudience/richaudience.go]: Done

File: adapters/richaudience/richaudience.go (hunk 2 above)
Comment (VeronikaSolovei9) [reply adapters/richaudience/richaudience.go:182]: There are no issues with this. We discussed this with the team and decided this is the right way to do this, similar to how we access Bid by index. I can revert it back.

**Review by SyntaxNode — COMMENTED** [review]
`request.Device.IP` also has a shared memory violation issue. @guscarreon these lines of code are covered by tests, why did the shared memory assertions pass here?

**Comment by bsardo** [comment]
> `request.Device.IP` also has a shared memory violation issue. @guscarreon these lines of code are covered by tests, why did the shared memory assertions pass here?

This does not appear to be an infrastructure issue; it's the tests. Let's update the `mockBidRequest` objects in all of the JSON tests by removing the following line `"keywords": "tagid=",` in the site and app objects wherever it is present. This line was masking the shared memory error since the `keywords` value computed by the adapter `MakeRequests` function was always the same as what was in the original bid request.

We should make sure that similar adjustments are made to the tests to detect shared memory issues with the device object.

**Comment by VeronikaSolovei9** [comment]
> request.Device.IP also has a shared memory violation issue. @guscarreon these lines of code are covered by tests, why did the shared memory assertions pass here?

Good catch about Device.Ip!
The reason why tests didn't catch it is because all tests have input `Device.ip` set to `"ip": "11.222.33.44",`.  To reproduce this problem input Device.ip should have been different from "11.222.33.44".

**Comment by VeronikaSolovei9** [comment]
> request.Device.IP also has a shared memory violation issue. @guscarreon these lines of code are covered by tests, why did the shared memory assertions pass here?

Good catch about Device.Ip!
The reason why tests didn't catch it is because all tests have input Device.ip set to "ip": "11.222.33.44". To reproduce this problem input Device.ip should be different from "11.222.33.44". Similar to `site|app.keywords`

**Review by bsardo — DISMISSED** [review]
LGTM

**Comment by VeronikaSolovei9** [comment]
@richaudience Can you please confirm (or not) you always expect only 1 impression per request? 
With recently added changes if request has more than one imp, `site|app.keywords` will be overwritten with the value from the last imp in array. In addition to this `raiExt.Test` per imp can be different which leads to inconsistency in data that should be saved in `request.Device.IP`. 
Current solution in this PR will work correctly and expected (in my understanding) if there is ONLY one impression in request.
Good catch @guscarreon

**Review by bsardo — APPROVED** [review]

**Review by guscarreon — APPROVED** [review]
LGTM
